docs/docsite/rst/dev_guide/testing_units_modules.rst

```diff
@@ -0,0 +1,534 @@
+dictionary for efficiency. This problem can be reduced somewhat with clear test naming so
+that the future developer immediately knows to delete the test case, but it's probably
+better to simply leave out the test case altogether and test for a real valuable feature
+of the code, such as installing all of the packages supplied as arguments to the module. 
+
+
+How to unit test Ansible modules
+================================
+
+There are a number of techniques which let you unit test modules.  Beware that most
+modules without unit tests are structured in a way that makes testing quite difficult and
+lead to very complicated tests which need more work than the code.  Effectively using unit
+tests may lead you to restructure your code. This is often a good thing and leads
+to better code overall. Good restructuring will make your code into clearer and more
+easily understood functions.
+
+
+Naming unit tests
+-----------------
+
+Unit tests should have logical names. If a developer working on the module being tested
+breaks the test case, it should be easy to figure what the unit test covers from the name.
+If a unit test is designed to verify compatibility with a specific software or API version
+then include the version in the name of the unit test.
+
+As an example ``test_v2_state_present_should_call_create_server_with_name()`` would be a
+good name, ``test_create_server()`` would not be.  
+
+
+Use of Mocks
+------------
+
+Mock objects (from https://docs.python.org/3/library/unittest.mock.html) can be very
+useful in building unit tests for special / difficult cases, but they can also
+lead to complex and confusing coding situations.  One good use for mocks would be in
+simulating an API. As for 'six', the 'mock' python package is bundled with Ansible (use
+'import ansible.compat.tests.mock'). See for example
+
+ensuring failure cases are visible with mock objects
+----------------------------------------------------
+
+Functions like module.fail_json() are normally expected to terminate execution. When you 
+run with a mock module object this doesn't happen since the mock always returns another mock 
+from a function call. You can set up the mock to raise an exception as shown above, or you can
+assert that these functions have not been called in each test. For example::
+
+  module = MagicMock()
+  function_to_test(module, argument)
+  module.fail_json.assert_not_called() 
+
+This applies not only to calling the main module but almost any other
+function in a module which gets the module object.  
+
+
+mocking of the actual module
+----------------------------
+
+The setup of an actual module is quite complex (see `Passing Arguments`_ below) and often
+isn't needed for most functions which use a module.  Instead you can use a mock object as
+the module and create any module attributes needed by the function you are testing.  If
+you do this then beware that the module::
```

Comment:
This appears to be an incomplete sentence:

... beware that the module {something missing here}?